### PR TITLE
Add stubs for flask-cors

### PR DESCRIPTION
**Pull request by kasium**

### CODE REVIEW

**Comment by kasium** [comment]
This package has dependencies to werkzeug and flask which are not part of typeshed. I currently used `Any` till #5768 is done

File: stubs/Flask-Cors/@tests/stubtest_allowlist.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+flask_cors.CORS.__init__
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/@tests/stubtest_allowlist.txt:1]: Could you comment (in this file) on why we need these allowlists?

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from logging import Logger
+from typing import Any, Iterable, Pattern, TypeVar
+
+_T = TypeVar("_T")
+_App = Any  # flask is not part of typeshed
+_Response = Any  # flask is not part of typeshed
+_Options = dict[str, Any]
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:7]: This could probably be a TypedDict.

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ACL_MAX_AGE: str
+ACL_REQUEST_METHOD: str
+ACL_REQUEST_HEADERS: str
+ALL_METHODS: list[str]
+CONFIG_OPTIONS: list[str]
+FLASK_CORS_EVALUATED: str
+RegexObject = Pattern[str]
+DEFAULT_OPTIONS: _Options
+
+def parse_resources(resources: dict[str, _Options] | list[str] | str) -> list[tuple[str, _Options]]: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:25]: The argument can also be a Pattern or any Iterable.

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+def get_regexp_pattern(regexp: str | Pattern[str]) -> str: ...
+def get_cors_origins(options: _Options, request_origin: str | None) -> list[str] | None: ...
+def get_allow_headers(options: _Options, acl_request_headers: str | None) -> str | None: ...
+def get_cors_headers(options: _Options, request_headers: dict[str, Any], request_method: str) -> _MultiDict: ...
+def set_cors_headers(resp: _Response, options: _Options) -> _Response: ...
+def probably_regex(maybe_regex: str | Pattern[str]) -> str: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:31]:
```suggestion
def probably_regex(maybe_regex: str | Pattern[str]) -> bool: ...
```

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+def re_fix(reg: str) -> str: ...
+def try_match_any(inst: str, patterns: Iterable[str]) -> bool: ...
+def try_match(request_origin: str, maybe_regex: str) -> bool: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:34]: Both of these accept re.Pattern for the second argument.

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+def get_cors_options(appInstance: _App | None, *dicts: _Options) -> _Options: ...
+def get_app_kwarg_dict(appInstance: _App | None = ...) -> _Options: ...
+def flexible_str(obj: Any) -> str: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:37]:
```suggestion
def flexible_str(obj: object) -> str | None: ...
```
It returns None for None. Use `object` to indicate that it accepts all types.

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+def serialize_option(options_dict: _Options, key: str, upper: bool = ...) -> None: ...
+def ensure_iterable(inst: Iterable[_T]) -> Iterable[_T]: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:39]: This accepts non-iterables. Giving it the right generic type might require overloads, but not sure if that's worth it.

File: stubs/Flask-Cors/flask_cors/decorator.pyi [hunk 8]
```diff
@@ -0,0 +1,17 @@
+from datetime import timedelta
+from logging import Logger
+from typing import Any, Callable
+
+LOG: Logger
+
+def cross_origin(
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/decorator.pyi:7]:
```suggestion
def cross_origin(
     *args: Any,
```
It looks like it ignores all *args. Meaningful arguments can only be passed as **kwargs.

File: stubs/Flask-Cors/flask_cors/decorator.pyi [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    origins: str | list[str],
+    methods: str | list[str],
+    expose_headers: str | list[str],
+    allow_headers: str | list[str],
+    supports_credentials: bool,
+    max_age: timedelta | int | str | None,
+    send_wildcard: bool,
+    vary_header: bool,
+    automatic_options: bool,
+) -> Callable[..., Any]: ...
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/decorator.pyi:17]: You can return a precise type here with ParamSpec from PEP 612.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks!

File: stubs/Flask-Cors/flask_cors/core.pyi (hunk 7 above)
Comment (kasium) [reply stubs/Flask-Cors/flask_cors/core.pyi:39]: Not sure how to express this correctly. I guess it' then Any for the argument and return type

File: stubs/Flask-Cors/flask_cors/decorator.pyi (hunk 9 above)
Comment (kasium) [reply stubs/Flask-Cors/flask_cors/decorator.pyi:18]: How can I use this w/o knowing the function?

File: stubs/Flask-Cors/flask_cors/core.pyi (hunk 7 above)
Comment (srittau) [reply stubs/Flask-Cors/flask_cors/core.pyi:39]: I believe something like this should work (please verify):

```python
_IterableT = TypeVar("_IterableT", bound=Iterable[Any])

@overload
def ensure_iterable(inst: str) -> list[str]: ...
@overload
def ensure_iterable(inst: _IterableT) -> _IterableT: ...
@overload
def ensure_iterable(inst: _T) -> list[_T]: ...
```

It's possible that mypy complains about "overlapping return types", which need to be `# type: ignore`'d.

File: stubs/Flask-Cors/flask_cors/core.pyi (hunk 2 above)
Comment (kasium) [reply stubs/Flask-Cors/flask_cors/core.pyi:7]: So the options are the same as e.g. for CORS.__init__. Is there a way to avoid duplicate definitions?

**Comment by kasium** [comment]
@JelleZijlstra can you please take a look why the CI fails?

File: stubs/Flask-Cors/flask_cors/core.pyi (hunk 7 above)
Comment (kasium) [reply stubs/Flask-Cors/flask_cors/core.pyi:39]: forgot about overload... thanks

File: stubs/Flask-Cors/flask_cors/core.pyi (hunk 2 above)
Comment (JelleZijlstra) [reply stubs/Flask-Cors/flask_cors/core.pyi:8]: Unfortunately not.

File: stubs/Flask-Cors/flask_cors/decorator.pyi (hunk 9 above)
Comment (JelleZijlstra) [reply stubs/Flask-Cors/flask_cors/decorator.pyi:18]: Something like this:

```suggestion
) -> Callable[[Callable[_P, ???]], Callable[_P, Response]]: ...
```
where `_P = ParamSpec("_P")`. It's not obvious from looking at the code what the return types of the original and the decorated functions are, so you'll have to figure that out.

**Comment by JelleZijlstra** [comment]
> @JelleZijlstra can you please take a look why the CI fails?

Looks like CI is green now. I responded to the two conversations that are still open.

**Comment by kasium** [comment]
@JelleZijlstra CI complains with `error: flask_cors.core.DEFAULT_OPTIONS variable differs from runtime type builtins.dict[Any, Any]
`. Not sure how to fix this...

File: stubs/Flask-Cors/flask_cors/core.pyi [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+from logging import Logger
+from typing import Any, Iterable, Pattern, TypeVar, overload
+from typing_extensions import TypedDict
+
+_IterableT = TypeVar("_IterableT", bound=Iterable[Any])
+_T = TypeVar("_T")
+_App = Any  # flask is not part of typeshed
+_Response = Any  # flask is not part of typeshed
+_MultiDict = Any  # werkzeug is not part of typeshed
+_Options = TypedDict(
```
Comment (JelleZijlstra) [inline stubs/Flask-Cors/flask_cors/core.pyi:11]: You should use class syntax for TypedDict: `class _Options(TypedDict, tootal=False):`.

**Comment by JelleZijlstra** [comment]
> @JelleZijlstra CI complains with error: flask_cors.core.DEFAULT_OPTIONS variable differs from runtime type builtins.dict[Any, Any] . Not sure how to fix this...

That looks like a stubtest bug (cc @hauntsaninja). You can add it to the allowlist for now (I think you used the allowlist in another PR already?) and report it as a bug over at https://github.com/python/mypy.

**Comment by hauntsaninja** [comment]
I was wondering when this would come up :-) I can try to fix soon

**Comment by kasium** [comment]
@JelleZijlstra done

**Pull request merged**